src/topp/GNPSExport.cpp

```diff
@@ -42,334 +59,523 @@
 
-          const BinnedSpectrum binned_highest_int(ms_maps[peptides[0].second.first][peptides[0].second.second], BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
+        MSExperiment merged_exp;
 
-          for (const auto& peptide : peptides)
-          {
-            int map_index = peptide.second.first;
-            int spectra_index = peptide.second.second;
-            // int highest_binned_intensity = peptide.first.first;
-            // auto highest_peptide_identification = peptide.first.second;
-
-            auto spectrum = ms_maps[map_index][spectra_index];
+        for (pair<int,int> &pept : pepts)
+        {
+          int map_index = pept.first;
+          int spec_index = pept.second;
+          auto test_spec = specs_list[msmaps_cached[map_index]][spec_index];
 
-            const BinnedSpectrum binned_spectrum(spectrum, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
+          BinnedSpectrum binned_spectrum(test_spec, bin_width, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
 
-            BinnedSpectralContrastAngle bsca;
-            double cosine_sim = bsca(binned_highest_int, binned_spectrum);
-            // OPENMS_LOG_DEBUG << cosine_sim << " >= " << cos_sim << endl;
+          BinnedSpectralContrastAngle bsca;
+          double cos_sim = bsca(binned_highest_int, binned_spectrum);
 
-            // compare calculated cosine sim to binned highest int
-            if (cosine_sim >= cos_sim)
-            {
-              for (Size spectrum_index = 0; spectrum_index < spectrum.size(); ++spectrum_index)
-              {
-                // exp.addSpectrum(spectrum);
-                // exp.addSpectrum(spectrum);
-                auto curr_spectrum = spectrum[spectrum_index];
-                if (ms2_block[curr_spectrum.getMZ()] < curr_spectrum.getIntensity())
-                {
-                  ms2_block[curr_spectrum.getMZ()] = curr_spectrum.getIntensity();
-                }
-              }
-            }
+          if (cos_sim >= cos_sim_threshold)
+          {
+            merged_exp.addSpectrum(test_spec);
           }
+        }
 
-
-          feature_stream << "BEGIN IONS" << endl;
-
-          feature_stream << "FEATURE_ID=" << feature_count++ << endl;
-          feature_stream << "CONSENSUSID=e_" << feature.getUniqueId() << endl;
-
-          feature_stream << "MSLEVEL=2" << endl;
-          feature_stream << "CHARGE=" << std::to_string(charge == 0 ? 1 : charge) << "+" << endl;
-          feature_stream << "PEPMASS=" << peptides[0].first.second.getMZ() << endl;
-          feature_stream << "FILE_INDEX=" << peptides[0].second.second << endl;
-          feature_stream << "RTINSECOND=" << peptides[0].first.second.getRT() << endl;
-
-          for (auto ms2_iter = ms2_block.rbegin(); ms2_iter != ms2_block.rend(); ++ms2_iter)
+        if (output_file.is_open())
```

Comment:
I feel that code redundancies of the output_types branches could be at least somewhat reduced to improve maintenance/readability.

E.g. by moving the case logic to only the specific, really different parts, or maybe adapting the 'merged' part to support filtered pepts that contain only the most intense one.